core/src/main/java/org/apache/iceberg/SnapshotProducer.java

```diff
@@ -215,7 +219,18 @@ public Snapshot apply() {
     long sequenceNumber = base.nextSequenceNumber();
     Long parentSnapshotId = parentSnapshot == null ? null : parentSnapshot.snapshotId();
 
-    validate(base, parentSnapshot);
+    try {
+      validate(base, parentSnapshot);
+    } catch (ValidationException ve) {
+      parentSnapshotId = snapshotIdToRollbackToOnConflict(parentSnapshotId);
```

Comment:
I think the logic that makes more sense in my mind is:

```
try {
  validate(base, parentSnapshot);
} catch (ValidationException ve) {
  if (shouldRollbackReplaceOnConflict()) {
     // do rollback logic 
 } else {
   // rethrow exception
 }
}
```

instead of having the logic to determine the flag as a part of `snapshotIdToRollbackToOnConflict` which is already determining the new parent snapshot.

What do you think?

Reply:
sounds good, made the changes